File: backend/api/system_config.py
"""Routes d'administration de la configuration système."""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel, Field

from backend.api.auth import get_current_user
from backend.core import models
from backend.core.system_config import (
    SystemConfig,
    get_config,
    rebuild_cors_middleware,
    save_config,
)

router = APIRouter()


class PublicSystemConfig(BaseModel):
    backend_url: str | None = Field(default=None)
    backend_url_lan: str | None = Field(default=None)
    backend_url_public: str | None = Field(default=None)
    frontend_url: str
    network_mode: str
    idle_logout_minutes: int = Field(60, ge=0, le=1440)
    logout_on_close: bool = False


@router.get("/public-config", response_model=PublicSystemConfig)
async def read_public_config() -> PublicSystemConfig:
    config = get_config()
    return PublicSystemConfig(
        backend_url=str(config.backend_url) if config.backend_url else None,
        backend_url_lan=str(config.backend_url_lan) if config.backend_url_lan else None,
        backend_url_public=str(config.backend_url_public) if config.backend_url_public else None,
        frontend_url=str(config.frontend_url),
        network_mode=config.network_mode,
        idle_logout_minutes=config.security.idle_logout_minutes,
        logout_on_close=config.security.logout_on_close,
    )


@router.get("/config", response_model=SystemConfig)
async def read_system_config(user: models.User = Depends(get_current_user)) -> SystemConfig:
    if user.role != "admin":
        raise HTTPException(status_code=403, detail="Autorisations insuffisantes")
    return get_config()


@router.post("/config", response_model=SystemConfig)
async def update_system_config(
    payload: SystemConfig,
    request: Request,
    user: models.User = Depends(get_current_user),
) -> SystemConfig:
    if user.role != "admin":
        raise HTTPException(status_code=403, detail="Autorisations insuffisantes")
    config = save_config(payload)
    rebuild_cors_middleware(request.app)
    return config

File: frontend/src/lib/systemConfig.ts
import { api } from "./api";
import { resetApiConfigCache } from "./apiConfig";

export type NetworkMode = "auto" | "lan" | "public";

export interface SystemConfig {
  backend_url?: string | null;
  backend_url_lan?: string | null;
  backend_url_public?: string | null;
  frontend_url: string;
  backend_host: string;
  backend_port: number;
  frontend_host: string;
  frontend_port: number;
  cors_origins: string[];
  network_mode: NetworkMode;
  extra: Record<string, string>;
  backend_public_url?: string | null;
}

export interface PublicSystemConfig {
  idle_logout_minutes: number;
}

function normalizeConfig(payload: SystemConfig): SystemConfig {
  const network_mode: NetworkMode =
    payload.network_mode === "lan" || payload.network_mode === "public" || payload.network_mode === "auto"
      ? payload.network_mode
      : payload.network_mode === "internet"
      ? "public"
      : "auto";

  return {
    ...payload,
    backend_url_public: payload.backend_url_public ?? payload.backend_public_url ?? payload.backend_url ?? null,
    backend_url_lan: payload.backend_url_lan ?? payload.backend_url ?? null,
    network_mode
  };
}

export async function fetchSystemConfig(): Promise<SystemConfig> {
  const response = await api.get<SystemConfig>("/system/config");
  return normalizeConfig(response.data);
}

export async function updateSystemConfig(payload: SystemConfig): Promise<SystemConfig> {
  const response = await api.post<SystemConfig>("/system/config", payload);
  resetApiConfigCache();
  return normalizeConfig(response.data);
}

export async function fetchPublicSystemConfig(): Promise<PublicSystemConfig> {
  const response = await api.get<PublicSystemConfig>("/system/public-config");
  return response.data;
}
